### PR TITLE
Backport support for nested optional, maps and tuples

### PR DESCRIPTION
**Pull request by ndcgundlach**
Backport support for nested optional, maps and tuples
## Change Description

This PR adds support for nested optional, maps and tuples in Mandel contracts.

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: tools/include/eosio/abigen.hpp [hunk 1]
```diff
@@ -163,6 +163,28 @@ namespace eosio { namespace cdt {
...
+            return name.substr(0,i+1);
+         };
+         kv.name = remove_ending_brackets(name);
+         kv.fields.push_back( {"key", get_template_argument_as_string(type)} );
+         kv.fields.push_back( {"value", get_template_argument_as_string(type, 1)} );
+         add_type(std::get<clang::QualType>(get_template_argument(type)));
+         add_type(std::get<clang::QualType>(get_template_argument(type, 1)));
+         _abi.structs.insert(kv);
+      }
+
```
Comment (ClaytonCalabrese) [inline tools/include/eosio/abigen.hpp]: Is this KV part that needs to be removed?

**Comment by ClaytonCalabrese** [comment]
Seems like toolchain_tests are failing as of this PR

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]

**Comment by ndcgundlach** [comment]
> Seems like toolchain_tests are failing as of this PR

I pushed up a fix for that.  You may want to retry.

**Comment by ClaytonCalabrese** [comment]
> > Seems like toolchain_tests are failing as of this PR
> 
> I pushed up a fix for that. You may want to retry.

You're right, my code was out of date, and it passes now.

**Review by ClaytonCalabrese — APPROVED** [review]

**Pull request merged**